[Master Bug 11513]- added OutOfOffice for the previuse owners

s7design opened the pull request:
Hi @mgruner 

Working on the bug #11513 ( http://bugs.otrs.org/show_bug.cgi?id=11513 ) we thought that it could be useful if there were OutOfOffice messages for the previus owners as wel as there are for the new owners in AgentTicketOwner screen.

If you think that this is needed feature you can merge it.

Regards
Zoran

carlosfrodriguez commented:
Hi @s7design,

That you for your pull request, but it seams to be not complete as for example AgentTciektMove will get the same problem, can you please track all places where this is needed to add and update this pull request?

Thank you in advance!

((enjoy))
cr

s7design commented:
Hi @carlosfrodriguez 

Yes you are right I got it now. I will check all screens and provide OutOdOffice message for them if it is needed.
Thanks for suggestion.

Regards
Zoran

s7design commented:
Hi @carlosfrodriguez,

I updated this PR, now there is OutOfOfice in the move screen as well.
Please let me know if I can improve ti.

Regards
Zoran

carlosfrodriguez commented:
Hi @s7design,

I think there is a lot of code duplication here, Ticket::TicketOwnerList() actually do a GetUserData() that does a GetPreferences(), so the information is already present and it does not need to be recalculated.

Please take a look on the information the is returned by Ticket::TicketOwnerList(), 'OutOfOfficeMessage' should be already there ;-)

This will made the fix much more simple

((enjoy))
cr

s7design commented:
Hi @carlosfrodriguez 

I agree with you. Thanks for suggestion. I updated PR. I believe  it is better now. I saw some unnecessary code and I had changed it as well.
Please let me know what do you think about that.

Regards
Zoran

carlosfrodriguez commented:
Hi @s7design 

This is not exactly what i had in mind, I think the solution at least for the TicketAxctionCommkon is just to add the OutOfOffice message, please check my comments in the commit

carlosfrodriguez commented on Kernel/Modules/AgentTicketActionCommon.pm:
```diff
@@ -1658,50 +1653,6 @@ sub _Mask {
-            ReturnSubType => 'OldOwner',
-            Data          => \%OldOwnersShown,
-            UserID        => $Self->{UserID},
-        );
```
Why remove all this stuff, ACLs might be needed here!

s7design replied:
I we use  _GetOldOwners there is the check for ACL 

    # workflow
    my $ACL = $TicketObject->TicketAcl(
        %Param,
        ReturnType    => 'Ticket',
        ReturnSubType => 'OldOwner',
        Data          => \%UserHash,
        UserID        => $Self->{UserID},
    );

carlosfrodriguez commented on Kernel/System/Ticket.pm:
```diff
@@ -4586,6 +4586,33 @@ sub TicketOwnerList {
+                $User{OutOfOfficeMessage} = "*** out of office till $TillDate/$Till d ***";
+            }
+        }
+
```
I think all of this is not really  needed, it should be already calculated

carlosfrodriguez commented on Kernel/Modules/AgentTicketActionCommon.pm:
```diff
@@ -1658,50 +1653,6 @@ sub _Mask {
-                };
-                $OldOwnersShown{$Key} = $Value;
-                $Counter++;
-            }
```
This is also important, as the ticket might have the same owner on multiple times

s7design replied:
I think it is not nedded if we use _GetOldOwners
There is code for that: 
next USER if $UserHash{ $User->{UserID} };

s7design commented:
Hi @carlosfrodriguez 

I updated PR. Thanks for your suggestion. I wanted to improve this code because there is some unneeded code. There is function _GetOldOwners, but it is was not used before.
We can speak about that more, but my opinion that there is needed cleaning code.

I believe we will find the best solution, feel free to contact me.

Regards
Zoran

carlosgarciac commented:
Hi @s7design, 

Could you please rebase this PR?, Then I could check it in a better way. Thx.


Best regards,
Carlos García

s7design commented:
Hi @carlosgarciac 

I rebased it, I think it is OK now.

Regards
Zoran

s7design commented:
Hi @carlosfrodriguez

I rebased this PR to make it up to date with the changes which have been happened in meantime.

Regards
Zoran

carlosfrodriguez commented on Kernel/Modules/AgentTicketActionCommon.pm:
```diff
@@ -1676,9 +1676,39 @@ sub _Mask {
+                my %Preferences = $UserObject->GetPreferences( UserID => $User->{UserID} );
+
+                # get time object
+                my $TimeObject = $Kernel::OM->Get('Kernel::System::Time');
```
Hi @s7design 

The Time object in OTRS 6 is deprecated, instead use DateTime

carlosfrodriguez commented on Kernel/Modules/AgentTicketActionCommon.pm:
```diff
@@ -1676,9 +1676,39 @@ sub _Mask {
+                    my $Time = $TimeObject->SystemTime();
+                    my $Start
+                        = "$Preferences{OutOfOfficeStartYear}-$Preferences{OutOfOfficeStartMonth}-$Preferences{OutOfOfficeStartDay} 00:00:00";
+                    my $TimeStart = $TimeObject->TimeStamp2SystemTime(
```
The idea of DateTime objects is that you can create multiple instances with defined dates and compare directly instead of converting to numbers

carlosfrodriguez commented on Kernel/Modules/AgentTicketMove.pm:
```diff
@@ -1220,8 +1220,7 @@ sub AgentMove {
                 Values => $Self->_GetOldOwners(
-                    QueueID  => $Param{DestQueueID},
-                    AllUsers => $Param{OwnerAll},
+                    QueueID => $Param{DestQueueID},
```
Why AllUsers was removed?

s7design replied:
Hi @carlosfrodriguez 

I deleted this param becouse it is not neccesary, it is used for example in the function _GetUsers but ont in _GetOldOwners. This param is no sense in _GetOldOwners becouse there is only old users no all users.

https://github.com/OTRS/otrs/blob/master/Kernel/Modules/AgentTicketMove.pm#L1549

Regards
Zoran

s7design commented:
Hi @carlosfrodriguez 

I implemented solution with the DataTime object, I believe it is much better now. Please let me know if you have any suggestion.

P.S. There are many places where deprecated Time object was used. I suppose it should be changed in further work. We are ready to help about that if it is necessary. 

Regards
Zoran

carlosfrodriguez commented on Kernel/Modules/AgentTicketActionCommon.pm:
```diff
@@ -1680,29 +1680,51 @@ sub _Mask {
-                        $OutOfOfficeMessage = "*** out of office till $TillDate/$Till d ***";
+
+                    if (   $TimeStart->Compare( DateTimeObject => $Time ) == -1
+                        && $TimeEnd->Compare( DateTimeObject => $Time ) == 1 )
```
Hi @s7design,

This could be actually $TimeStart > $Time && TimeEnd > $Time, since DateTime object supports operator overload, have you tried?

carlosfrodriguez commented on Kernel/System/User.pm:
```diff
@@ -266,18 +266,31 @@ sub GetUserData {
+
+            if (   $TimeStart->Compare( DateTimeObject => $Time ) == -1
+                && $TimeEnd->Compare( DateTimeObject => $Time ) == 1 )
+            {
```
Hi @s7design,

This could be  $TimeStart > $Time && TimeEnd > $Time, since DateTime too.

s7design replied:
Oh thanks, I will try, it sounds great if it is possible

s7design commented:
Hi @carlosfrodriguez 

Thanks for the follow-up, we will add the test. 

Regards
Zoran

s7design commented:
Hi @carlosfrodriguez 

I updated this PR and implemented in the code suggestion that you mentioned in the comments. The Unit test is extended, I am still working on the Selenium test. 

I am trying to find out how to handle with the modernize fields and "Previous owner" feature in them. I think you and I have already spoken about UX with such previous owner ( three click to redraw field with previous owners ). Now I have a problem how I open modernize field, select Filter button and then enable check box "Previous  owner". I manged to do that in the browser console but the same jQuery in Selenium test did not work when I use $Selenium->execute_script(). 
I write all of this just to inform you, but I believe I will succeed to find any solution to test this feature. 

Regards
Zoran

s7design commented:
Hi @carlosfrodriguez 

In the last commit there is extended Selenium test. Hereby is tested Out of office message for new owner and the previous ones in the screen AgentTicketOwner.

I'm not completely satisfied how the test looks like, but I did not mange to find better solution in this moment.

For example: I should use thwo focus click here:

```
       # expand 'New owner' input field
        $Selenium->execute_script("\$('#NewOwnerID_Search').focus().focus()");
```

If you know how we can improve it please let me know.

Regards
Zoran

s7design commented:
Hi @dvuckovic 

This PR is rebased

Regards
Zoran

s7design commented:
Hello,

Follow up fix has been applied. Closing this PR.

Regards,
Sanjin
